### PR TITLE
V6 es6: RF query implementation

### PR DESCRIPTION
**Pull request by hailem**

### CODE REVIEW

File: src/query.ts [hunk 1]
```diff
@@ -75,6 +90,16 @@ export class Query {
     return this._getFilter(this._excludedFilters);
   }
 
+  public get sort(): GenericObject {
+    const result: InputExtendedSortValue[] = [];
+    this._sort.forEach((value, key) => {
```
Comment (carlosescri) [inline src/query.ts]: Rename: `value => order`, `key => field`

File: src/query.ts [hunk 2]
```diff
@@ -75,6 +90,16 @@ export class Query {
...
+      const sortValue: GenericObject = {};
```
Comment (carlosescri) [inline src/query.ts]: Seems that this should be `InputExtendedSortValue` instead of `GenericObject`.

File: src/query.ts [hunk 3]
```diff
@@ -75,6 +90,16 @@ export class Query {
...
+      sortValue[key] = value;
+      result.push(sortValue);
```
Comment (carlosescri) [inline src/query.ts]: Probably you could do:

```js
this._sort.forEach((ordering, field) => results.push({[field]: ordering}))
```

File: src/query.ts [hunk 4]
```diff
@@ -138,7 +163,7 @@ export class Query {
    * @param filterName  The name of the filter to modify
    * @param value       The value to remove from the filter
    */
-  public removeExcludedFilter(filterName: string, value: AssignFilterValue): Query {
+  public removeExcludedFilter(filterName: string, value: InputFilterValue): Query {
```
Comment (carlosescri) [inline src/query.ts]: Rename this one: `removeExcludedFilter => removeExcludeFilter` or rename `addExcludeFilter => addExcludedFilter` (`Exclude => Excluded`)

File: src/query.ts [hunk 5]
```diff
@@ -244,77 +252,32 @@ export class Query {
...
-   * Remove a sorting field from the parameters
-   *
-   * @param   {String}    field   The field we want to remove the ordering
-   *
-   */
-  public removeSorting(field: string): void {
-    if (!this.params.sort) {
-      return;
+    if ('sort' in this.params) {
+      this.addSort(this.params['sort']);
```
Comment (carlosescri) [inline src/query.ts]: Remove `sort` key from `this.params` after importing sort parameters.

File: src/query.ts (hunk 1 above)
Comment (hailem) [reply src/query.ts]: done

File: src/query.ts (hunk 2 above)
Comment (hailem) [reply src/query.ts]: done

File: src/query.ts (hunk 3 above)
Comment (hailem) [reply src/query.ts]: done

File: src/query.ts (hunk 4 above)
Comment (hailem) [reply src/query.ts]: done

File: src/query.ts (hunk 5 above)
Comment (hailem) [reply src/query.ts]: done

File: src/types/request.ts [hunk 6]
```diff
@@ -59,8 +49,8 @@ export interface DoofinderParameters {
   type?: string | string[];
   filter?: RequestFiltersObject;
   exclude?: RequestFiltersObject;
-  transformer?: TransformerOptions;
-  sort?: RequestSortOptions;
+  transformer?: any;
```
Comment (carlosescri) [inline src/types/request.ts]: Use `string` here.

File: src/types/request.ts [hunk 7]
```diff
@@ -59,8 +49,8 @@ export interface DoofinderParameters {
...
+  sort?: any;
```
Comment (carlosescri) [inline src/types/request.ts]: Don't use `any` here.

File: src/query.ts [hunk 8]
```diff
@@ -204,231 +243,174 @@ export class Query {
...
-   */
-  public removeType(type: string): void {
-    if (this.params.type) {
-      if (isArray(this.params.type)) {
-        const index: number = this.params.type.indexOf(type);
-        if (index !== -1) {
-          (this.params.type as Array<unknown>).splice(1, index);
+  public addTypes(types?: string | string[]): Query {
+    if (types) {
+      this._dataTypes = new Set();
```
Comment (carlosescri) [inline src/query.ts]: `addTypes => types()`
`_dataTypes => _types` is good enough and shorter…

File: src/query.ts [hunk 9]
```diff
@@ -204,231 +243,174 @@ export class Query {
...
   /**
    * Sets the transformer, call it empty to reset it to null
    *
-   * @param  {String}   transformer   The transformer option to set
+   * @param  transformer - The transformer option to set
+   *
+   * @returns Query
    */
-  public transformer(transformer?: TransformerOptions): void {
+  public transformer(transformer?: string): Query {
```
Comment (carlosescri) [inline src/query.ts]: This doesn't handle the case `query.transformer(null)`. And don't validate the parameter, just set it. This is a type of parameter we can customize for a user in the config and a validation won't let the user set it.

File: src/query.ts [hunk 10]
```diff
@@ -204,231 +243,174 @@ export class Query {
...
-    const index = this._findSortField(field, sortParams as Array<RequestSortOptions>);
-    if (index !== -1) {
-      (sortParams as Array<RequestSortOptions>).splice(index, 1);
-      if (sortParams.length === 0) {
-        delete this.params.sort;
-      } else {
-        this.params.sort = sortParams;
-      }
+    if ('rpp' in this.params) {
+      this.resultsPerPage(this.params['rpp']);
```
Comment (carlosescri) [inline src/query.ts]: I would name this `rpp` instead, it's much shorter and matches the parameter name.

File: src/query.ts [hunk 11]
```diff
@@ -204,231 +243,174 @@ export class Query {
...
+    return this;
   }
 
   /**
    * Advances the current page to the next one
    *
    */
   public nextPage(): void {
-    this.setParameter('page', (this.params.page || 1) + 1);
+    this._page = (this._page || 1) + 1;
```
Comment (carlosescri) [inline src/query.ts]: I would do `this._page++` and that's all. See my comment on `page(n)`.

File: src/query.ts [hunk 12]
```diff
@@ -204,231 +243,174 @@ export class Query {
...
-   * @param  {Number}     page              The page we want to set
+   * @param  page - The page we want to set
+   *
+   * @returns Query
    *
    */
-  public page(page?: number): void {
-    if (page) {
-      this.setParameter('page', page);
+  public page(page?: number): Query {
```
Comment (carlosescri) [inline src/query.ts]: In this case I would reset page to `1` if the provided value is `0`, `undefined` or `null`. This is a parameter with a clear default value.

```typescript
const value: number = parseInt(page || 1, 10);
if (isNaN(value)) {
  throw new QueryValueError();
} else {
  this._page = value;
}
```

File: src/query.ts [hunk 13]
```diff
@@ -204,231 +243,174 @@ export class Query {
...
    *
-   * @param  {Number}   rpp   The results per page to set
+   * @param  rpp - The results per page to set
    *
+   * @returns Query
    */
-  public resultsPerPage(rpp?: number): void {
-    if (rpp) {
-      this.setParameter('rpp', rpp);
+  public resultsPerPage(rpp?: number): Query {
```
Comment (carlosescri) [inline src/query.ts]: Rename to `rpp`. This is another special parameter. There's a forced max value of `100`, so the number shouldn't be greater than that. And we have a default value of… ¿`20`?

I think the default should be configurable via a static property, something like `Query.defaultRpp = 20`, being that the default.

So, you would do something like with `page` here:

```typescript
const value: number = parseInt(rpp || Query.defaultRpp);
if (isNaN(value)) {
  throw new QueryValueError();
} else {
  this._rpp = Math.min(value, MAX_RPP); // not configurable
}
```

File: src/query.ts [hunk 14]
```diff
@@ -204,231 +243,174 @@ export class Query {
...
    * Sets the timeout for the query, call it empty to reset
    *
-   * @param  {Number}   timeout       The timeoout for the call
+   * @param  timeout - The timeoout for the call
    *
    */
   public timeout(timeout?: number): void {
     if (timeout) {
-      this.setParameter('timeout', timeout);
+      this._timeout = timeout;
```
Comment (carlosescri) [inline src/query.ts]: Is there a default timeout in the service? Is it greater than `0`? If that's the case, this doesn't work when the passed value is `0`. For instance, you call `query.timeout(0)` and you get no timeout parameter.

Also, this uses `this._timeout` and `this.params.timeout` at the same time… 🤔

File: src/query.ts [hunk 15]
```diff
@@ -438,12 +420,12 @@ export class Query {
    * Allows to ask for jsonp format, call without parameters
    * to clear the flag
    *
-   * @param  {Boolean}    jsonp   Wether to use jsonp or not
+   * @param  jsonp - Wether to use jsonp or not
```
Comment (carlosescri) [inline src/query.ts]: `wether => whether`

File: src/query.ts [hunk 16]
```diff
@@ -438,12 +420,12 @@ export class Query {
...
    *
    */
   public jsonp(jsonp?: boolean): void {
     if (jsonp) {
-      this.setParameter('jsonp', jsonp);
+      this._jsonp = jsonp;
```
Comment (carlosescri) [inline src/query.ts]: This uses `this._jsonp` and `this.params.jsonp` at the same time!

File: src/query.ts [hunk 17]
```diff
@@ -456,26 +438,36 @@ export class Query {
    * @param  {String}   queryName   The query_name parameter value to set
    *
    */
-  public queryName(queryName?: QueryTypes): void {
-    if (queryName) {
-      this.setParameter('query_name', queryName);
+  public queryName(queryName?: string): Query {
```
Comment (carlosescri) [inline src/query.ts]: Any non-string value will be _coerced_ to a string in the url querystring… so I think this validation is not necessary.

Also, don't _delete_ `this._queryName` (you're destroying the property in the instance), just set it to undefined.

These kind of checks with no distinction between `null` and `undefined` could be:

```typescript
if (queryName == null || `${queryName}`.trim().length === 0) {
  this._queryName = undefined;
} else if (typeof queryName === 'string') {
  this._queryName = queryName;
} else {
  throw…
}
```

File: src/query.ts [hunk 18]
```diff
@@ -456,26 +438,36 @@ export class Query {
...
    * Sets the nostats flag, call without parameters to clear it
    *
-   * @param  {Boolean}    nostats   Wether to send the nostats flag or not
+   * @param - nostats   Wether to send the nostats flag or not
    *
    */
-  public noStats(nostats?: boolean): void {
-    if (nostats && !!nostats) {
-      this.setParameter('nostats', 1);
+  public noStats(noStats?: boolean | number): Query {
```
Comment (carlosescri) [inline src/query.ts]: Don't allow numbers in the type definition if value must be a boolean. Just coerce to boolean unless the value is null or undefined. If the value must be a number in the query params, just coerce to boolean and use it to set the number.

```js
if (noStats == null) {
  this._noStats = undefined;
} else {
  this._noStats = !!noStats ? 1 : 0;
}
```

I wouldn't validate this more than that.

File: src/query.ts [hunk 19]
```diff
@@ -638,4 +617,24 @@ export class Query {
...
+    if (sortType.toLowerCase() === OrderType.ASC) {
+      return OrderType.ASC;
+    } else if (sortType.toLowerCase() === OrderType.DESC) {
+      return OrderType.DESC;
+    } else {
+      throw new QueryValueError('Value error: Sort type must be: "asc" or "desc"');
+    }
+  }
+
+  private _validateTransformer(transformer: string): TransformerOptions {
```
Comment (carlosescri) [inline src/query.ts]: For me this is useless.

File: src/query.ts [hunk 20]
```diff
@@ -638,4 +617,24 @@ export class Query {
       (!value || isPlainObject(filter.get(filterName)) || (filter.get(filterName) as TermsFilter).has(value))
     );
   }
+
+  private _validateOrderType(sortType: string): OrderType {
```
Comment (carlosescri) [inline src/query.ts]: This is useful because there're only two possible values.

File: src/query.ts (hunk 11 above)
Comment (hailem) [reply src/query.ts]: this._page could be undefined

Comment (carlosescri) [reply src/query.ts]: You didn't read my comment about `this.page(n)`. The proposal makes impossible to have `undefined` in `this._page` unless explicitly set without using the public API, so it's not our problem.

File: src/query.ts (hunk 13 above)
Comment (hailem) [reply src/query.ts]: I can't do that because rpp must be a string in parseInt

Comment (carlosescri) [reply src/query.ts]: In JS you can pass anything, I don't know if TS will complain. For instance:

```js
parseInt({a: 23}, 10); // NaN
```

File: src/query.ts [hunk 21]
```diff
@@ -328,8 +305,8 @@ export class Query {
    *
    */
   public page(page?: number): Query {
-    if (typeof page === 'number' || typeof page === 'undefined') {
-      this._page = page;
+    if (typeof page === 'number' || typeof page === 'undefined' || typeof page === null) {
+      this._page = page && page > 0 ? page : 0;
```
Comment (carlosescri) [inline src/query.ts]: This should be `1` if the condition is not met.

File: src/query.ts [hunk 22]
```diff
@@ -393,11 +370,11 @@ export class Query {
...
    */
-  public transformer(transformer?: string): Query {
-    if (transformer) {
-      this._transformer = this._validateTransformer(transformer);
+  public transformer(transformer?: string | null): Query {
+    if (transformer || transformer === null) {
+      this._transformer = transformer;
     } else {
-      this._transformer = undefined;
+      delete this._transformer;
```
Comment (carlosescri) [inline src/query.ts]: Don't delete class attributes!

File: src/query.ts [hunk 23]
```diff
@@ -409,25 +386,25 @@ export class Query {
...
-   * @param  jsonp - Wether to use jsonp or not
+   * @param  jsonp - Whether to use jsonp or not
    *
    */
   public jsonp(jsonp?: boolean): void {
     if (jsonp) {
       this._jsonp = jsonp;
     } else {
-      delete this.params.jsonp;
+      delete this._jsonp;
```
Comment (carlosescri) [inline src/query.ts]: Don't delete class attributes!

File: src/query.ts [hunk 24]
```diff
@@ -440,11 +417,7 @@ export class Query {
...
   public queryName(queryName?: string): Query {
     if (typeof queryName === 'string' || typeof queryName === 'undefined') {
-      if (queryName) {
-        this._queryName = queryName;
-      } else {
-        delete this._queryName;
-      }
+      this._queryName = queryName;
     } else {
       throw new QueryValueError('Value error: queryname must be an string value');
```
Comment (carlosescri) [inline src/query.ts]: typo: `… a string …`

File: src/query.ts [hunk 25]
```diff
@@ -454,11 +427,11 @@ export class Query {
...
+   * @param - noStats   Wether to send the nostats flag or not
    *
    */
-  public noStats(noStats?: boolean | number): Query {
-    if (['boolean', 'string', 'undefined'].includes(typeof noStats)) {
+  public noStats(noStats?: boolean): Query {
+    if (['boolean', 'undefined'].includes(typeof noStats)) {
       if (noStats) {
         this._noStats = 1;
       } else {
```
Comment (carlosescri) [inline src/query.ts]: Don't delete class attributes!

**Pull request merged**